utils/trie_writer.hh

```diff
@@ -0,0 +1,1124 @@
+
+private:
+    using buf = std::unique_ptr<std::byte[]>;
+    // We keep the last freed segment here. If we need a new buffer and this is present, we reuse it instead of allocating a new one.
+    // This ensures we won't repeatedly free and allocate a segment from the main allocator.
+    buf _spare = nullptr;
+    // The segment currently at the top of the stack.
+    buf _current = nullptr;
+    // Remaining free space in _current.
+    size_t _remaining = 0;
+    // Previous segments in the stack.
+    std::vector<buf> _old;
+
+#if SANITIZE_BUMP_ALLOCATOR
+    // The number of discard() calls until now.
+    uint64_t _discard_timestamp = 0;
+    // Maps given stack positions to the index of the last discard() covering that position.
+    // A valid pointer to some position will have a _discard_timestamp which matches this map.
+    // An invalidated pointer will have a _discard_timestamp which is smaller than that.
+    std::map<size_t, size_t> _max_legal_timestamp{{0, 0}};
+#endif
+
+private:
+    // Returns the current total size of the "stack".
+    size_t global_pos() const {
+        return allocated_memory() - _remaining;
+    }
+
+    // Invalidate all bytes in the stack since the byte at given address (inclusive).
+    void discard_impl(std::byte* addr) noexcept {
+        EXPENSIVE_ASSERT(_current);
+        while (!(addr > _current.get() && addr <= _current.get() + _segment_size)) [[unlikely]] {
+            #if SANITIZE_BUMP_ALLOCATOR
+                std::memset(_current.get(), 0xfe, _segment_size);
+            #endif
+            _spare = std::move(_current);
+            EXPENSIVE_ASSERT(!_old.empty());
+            _current = std::move(_old.back());
+            _remaining = 0;
+            _old.pop_back();
+        }
+        #if SANITIZE_BUMP_ALLOCATOR
+            auto end = _current.get() + _segment_size - _remaining;
+            std::memset(addr, 0xfe, end - addr);
+        #endif
+        _remaining = _segment_size - (addr - _current.get());
+    }
+
+    void push_segment() {
+        if (_current) {
+            _old.push_back(std::move(_current));
+        }
+        if (_spare) {
+            _current = std::move(_spare);
+        } else {
+            _current = std::unique_ptr<std::byte[]>(
+                static_cast<std::byte*>(aligned_alloc(_segment_size, _segment_size)));
+        }
+    }
+
+public:
+    bump_allocator(size_t segment_size) : _segment_size(segment_size) {
+        SCYLLA_ASSERT(_segment_size % alignof(max_align_t) == 0);
+    }
+
+    // Total memory usage by this allocator.
+    size_t allocated_memory() const {
+        return (_old.size() + bool(_current)) * _segment_size;
+    }
+
+    // Allocate a suitably-aligned (uninitialized) buffer for an array of N elements of type T.
+    // The total size of this allocation must be smaller than segment_size.
+    template <typename T>
+    [[nodiscard]] ptr<T> alloc(size_t n) {
+        static_assert(alignof(T) <= segment_alignment);
```

Comment:
It would aid readability if the type we ptr\<T[]> or buf_ptr\<T>, to emphasize that it's not a pointer to one T.

Reply:
In v2, I modified it so that `ptr<T>` and `ptr<T[]>` can both be used. (Even though in my opinion that's overkill).